### PR TITLE
[Feat] Enable `header` and `footer` inside `Graph`, `Table` and `AgGrid`

### PR DESCRIPTION
**Pull request by huong-li-nguyen**
## Description
- Add warning if figure.title is used instead of graph.title
- Enable `header` and `footer` in `Graph`, `Table` and `AgGrid`
- Enable `title` in `Graph`
- Some quick design updates to `dash_data_table` to look more similar to `dash_ag_grid`
- Update `margin_b` in dashboard overwrites

**TO DO:**

- [x] Update existing unit tests
- [x] Unit test for warning
- [x] Unit tests for header / footer / title
- [x] Update dev example
- [x] Docs

@l0uden 
- [x] Ask Alexey to create a screenshot test
- [x] Update screenshot tests -> https://github.com/mckinsey/vizro-qa/pull/111 - @l0uden to approve

## Notice

- [x] I acknowledge and agree that, by checking this box and clicking "Submit Pull Request":

  - I submit this contribution under the [Apache 2.0 license](https://www.apache.org/licenses/LICENSE-2.0.txt) and represent that I am entitled to do so on behalf of myself, my employer, or relevant third parties, as applicable.
  - I certify that (a) this contribution is my original creation and / or (b) to the extent it is not my original creation, I am authorized to submit this contribution on behalf of the original creator(s) or their licensees.
  - I certify that the use of this contribution as authorized by the Apache 2.0 license does not violate the intellectual property rights of anyone else.
  - I have not referenced individuals, products or companies in any commits, directly or indirectly.
  - I have not added data or restricted code in any commits, directly or indirectly.

### CODE REVIEW

File: vizro-core/tests/unit/vizro/models/_components/test_table.py
Comment (huong-li-nguyen) [inline vizro-core/tests/unit/vizro/models/_components/test_table.py:149]: Remember we don't need to use the `id` everywhere. The `keys_to_strip` argument is exactly for this case. Here we want to use mandatory arguments, so no need to specify id. 

There is a separate test for testing the ID in this file as well.

File: vizro-core/src/vizro/models/_components/graph.py
Comment (huong-li-nguyen) [inline vizro-core/src/vizro/models/_components/graph.py:129]: For reviewers: If someone uses the title argument inside a px chart, we've decided to not try and align it with other components anymore. Only if they use the `Graph.title` it will be properly aligned. A warning is raised when they use the `title` argument inside the px.chart

File: vizro-core/src/vizro/models/_components/graph.py
Comment (maxschulz-COL) [inline vizro-core/src/vizro/models/_components/graph.py:47]:
```suggestion
        description="Markdown text positioned below the `Graph.title`, compliant with the CommonMark specification.",
```
Just to avoid confusion with what component is really meant

File: vizro-core/src/vizro/models/_components/ag_grid.py
Comment (maxschulz-COL) [inline vizro-core/src/vizro/models/_components/ag_grid.py:42]:
```suggestion
        description="Markdown text positioned below the `AgGrid.title`, compliant with the CommonMark specification.",
```

File: vizro-core/src/vizro/models/_components/graph.py
Comment (maxschulz-COL) [reply vizro-core/src/vizro/models/_components/graph.py:47]: @lingyielia : I am curious if the word `subtitle` in the description would really help the model to populate this field?

@huong-li-nguyen If it make no difference, would it be possible to weave in the work subtitle into the description. It would be useful for human readers of the docstring too 🤔 ? Like "Markdown text (e.g. a subtitle) ..."

File: vizro-core/src/vizro/models/_components/table.py
Comment (maxschulz-COL) [inline vizro-core/src/vizro/models/_components/table.py:41]:
```suggestion
        description="Markdown text positioned below the `Table.title`, compliant with the CommonMark specification.",
```

File: vizro-core/src/vizro/static/css/table.css
Comment (maxschulz-COL) [inline vizro-core/src/vizro/static/css/table.css:2]: Did this hurt 😂 ?

File: vizro-core/mkdocs.yml
Comment (maxschulz-COL) [inline vizro-core/mkdocs.yml:32]:
```suggestion
          - Title, header, footer: pages/user-guides/titles-header-footer.md
```
or all plural?

File: vizro-core/docs/pages/user-guides/titles-header-footer.md
Comment (maxschulz-COL) [inline vizro-core/docs/pages/user-guides/titles-header-footer.md:5]:
```suggestion
The [`Graph`][vizro.models.Graph], the [`Table`][vizro.models.Table] and the [`AgGrid`][vizro.models.AgGrid] models
```

File: vizro-core/docs/pages/user-guides/titles-header-footer.md
Comment (maxschulz-COL) [inline vizro-core/docs/pages/user-guides/titles-header-footer.md:11]:
```suggestion
- **footer**: Accepts Markdown text, commonly used for citing data sources, providing information on the last update, or adding disclaimers.
```

File: vizro-core/docs/pages/user-guides/titles-header-footer.md
Comment (maxschulz-COL) [inline vizro-core/docs/pages/user-guides/titles-header-footer.md:34]:
```suggestion
                    figure=px.scatter(df, x="sepal_width", y="sepal_length", color="species"),
```
Discovered actually with PyCafe :). The example still does not run, but will once we release this change.

That means that this specific cell won't run in the `latest` docs, but it will run after release. Do you think that's fine?

**Review by maxschulz-COL — APPROVED** [review]
Love this change, really nice to look at! Have a few comments, but approving pending the implementation of those :)

**Review by petar-qb — COMMENTED** [review]
I really like this initiative 😍.

This also makes initial page loading much prettier as the title is introduced for the Graph component as well (and will be even more prettier when we introduce some shinning div boxes for dynamic components - similar like it's done for [dash documentation](https://dash.plotly.com/)).

File: vizro-core/docs/pages/user-guides/titles-header-footer.md
Comment (maxschulz-COL) [reply vizro-core/docs/pages/user-guides/titles-header-footer.md:1]: This is something done at release, not part of this PR.

Comment (petar-qb) [reply vizro-core/docs/pages/user-guides/titles-header-footer.md:1]: I added a reminder note in the release process page for this case.

File: vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py
Comment (huong-li-nguyen) [reply vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py:162]: I think we are already aligned on this one already, because that's what I was intending to do, but you just made me notice that I forgot to remove i`d="underlying_ag_grid_id"` in some places :) 

So I've removed the `id` from the unit tests where it was not necessary so `test_XX_grid_build_mandatory_only` and`test_XXXX_build_title_header_footer`. However, there is a separate test called `test_XXX_build_with_underlying_id `in both tables already 👍

File: vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py
Comment (huong-li-nguyen) [reply vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py:213]: Just forgot to remove it here!

File: vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py
Comment (huong-li-nguyen) [reply vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py:153]: There is a test below called `test_ag_grid_build_with_underlying_id`

File: vizro-core/changelog.d/20240902_101331_huong_li_nguyen_enable_header_footer.md
Comment (huong-li-nguyen) [reply vizro-core/changelog.d/20240902_101331_huong_li_nguyen_enable_header_footer.md:28]: Hmm.. not sure if it's worth doing as there is a warning returned already. I can add it if you guys prefer it, but I am indifferent here 😄 

@petar-qb @maxschulz-COL

File: vizro-core/src/vizro/models/_components/graph.py
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_components/graph.py:1]: Antony and I have decided to hold off on that for now. Unlike the graph and tables, it's unclear whether the Figure actually requires these arguments. A good example are the KPI cards, which don't need a title, header, or footer since the card itself serves as a markdown.

If there is significant demand, we can always add it later, but I wouldn't include it from the beginning.

File: vizro-core/docs/pages/user-guides/titles-header-footer.md
Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/titles-header-footer.md]: Yes, I think that's fine 👍 Not sure how we would do it otherwise. Adding the examples in after the release doesn't really make sense, because then the feature is not really "ready".

File: vizro-core/src/vizro/static/css/table.css
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/static/css/table.css:2]: Yes 😅  I actually wished that we wouldn't have the DataTable in our code basis anymore

File: vizro-core/src/vizro/models/_components/graph.py
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_components/graph.py:47]: Yes, will add it in!

File: vizro-core/src/vizro/models/_components/graph.py
Comment (petar-qb) [reply vizro-core/src/vizro/models/_components/graph.py:1]: Fine by me 👍

File: vizro-core/src/vizro/models/_components/graph.py
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_components/graph.py:152]: I like that!

For context:

Screenshot 2024-09-04 at 12 44 31

File: vizro-core/changelog.d/20240902_101331_huong_li_nguyen_enable_header_footer.md
Comment (petar-qb) [reply vizro-core/changelog.d/20240902_101331_huong_li_nguyen_enable_header_footer.md:28]: I think all Vizro users somewhere in their app have a Graph.figure.title, and will probably try to find what's the reason behind the "strange" warnings after Vizro is updated. I guess that their next move will be searching for Vizro changelog file and trying to find out what's happening. So, as this will hit all (or almost all) Vizro users, I think we should mention this change to the changelog, even though we don't prefer mentioning minor styling changes.

Any strong opinion on this @maxschulz-COL?

File: vizro-core/docs/pages/user-guides/titles-header-footer.md
Comment (petar-qb) [reply vizro-core/docs/pages/user-guides/titles-header-footer.md]: @huong-li-nguyen Do you think we don't have to add: `import vizro.plotly.express as px` here? (and in the "Formatted DataTable" example too)

File: vizro-core/changelog.d/20240902_101331_huong_li_nguyen_enable_header_footer.md
Comment (huong-li-nguyen) [reply vizro-core/changelog.d/20240902_101331_huong_li_nguyen_enable_header_footer.md:28]: If you have a strong preference for it, let me just add it in :)

File: vizro-core/docs/pages/user-guides/title-header-footer.md
Comment (petar-qb) [inline vizro-core/docs/pages/user-guides/title-header-footer.md:3]: Is this okay? 
```suggestion
# How to add title, header and footer

This guide explains how to add title, header, and footer to your components in Vizro.
```

Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/title-header-footer.md:3]: Singular

File: vizro-core/docs/pages/user-guides/titles-header-footer.md
Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/titles-header-footer.md]: Ah I didn't get that 😆 I thought this was just adding a line before - will add it 👍

File: vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py
Comment (petar-qb) [reply vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py]: I like the latest version since it's quite clear and separate what's the `test_XX_grid_build_mandatory_only `, `test_XXXX_build_title_header_footer`, and `test_XXX_build_with_underlying_id`. 👍 

However, I think we should also test (in a separate test) a case when `vm.AgGrid.title` is defined, but the underlying id isn't. In that case, `vm.AgGrid.build` method should return `f"__input_{self.id}"` for the underlying `html.Div`.

Comment (petar-qb) [reply vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py]: So, something like this:
```py
   def test_ag_grid_build_with_id_and_without_underlying_id(self, standard_ag_grid, gapminder):
        ag_grid = vm.AgGrid(id="text_ag_grid", figure=standard_ag_grid)
        ag_grid.pre_build()
        ag_grid = ag_grid.build()

        expected_ag_grid = dcc.Loading(
            html.Div(
                [
                    None,
                    None,
                    html.Div(
                        id="text_ag_grid",
                        children=[html.Div(id="__input_text_ag_grid")],
                        className="table-container",
                    ),
                    None,
                ],
                className="figure-container",
            ),
            color="grey",
            parent_className="loading-container",
            overlay_style={"visibility": "visible", "opacity": 0.3},
        )
        assert_component_equal(ag_grid, expected_ag_grid)
```

Comment (petar-qb) [reply vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py]: Do you also mind to remove the `filter_interaction_action` from:
 - `test_pre_build_actions_underlying_ag_grid_id` and 
 - `test_ag_grid_build_with_underlying_id` 

in `test_ag_grid` and `test_table` files?

File: vizro-core/changelog.d/20240902_101331_huong_li_nguyen_enable_header_footer.md
Comment (maxschulz-COL) [reply vizro-core/changelog.d/20240902_101331_huong_li_nguyen_enable_header_footer.md:26]: Lets do it!

File: vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py
Comment (huong-li-nguyen) [reply vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py]: Ahh, I see. So the `test_ag_grid_build_mandatory_only` was previously used for this purpose as well. That wasn't clear to me. In this case, I'll add above in and I think having it separate makes it also clearer what that is actually testing!

Comment (huong-li-nguyen) [reply vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py]: I've parameterized the test :) Will push in a minute

File: vizro-core/src/vizro/models/_components/graph.py
Comment (huong-li-nguyen) [reply vizro-core/src/vizro/models/_components/graph.py]: Using the triple quotes actually makes it a bit more difficult to read (as an output), because the formatting seems off then 🤔

**Triple quotes**
Screenshot 2024-09-04 at 14 18 48

**Single quotes**
Screenshot 2024-09-04 at 14 22 33

Took me forever to figure out why the unit tests were failing with this change 😄 But if you use parentheses, you have to change the syntax of match statement: https://github.com/pytest-dev/pytest/issues/10595

**Comment by huong-li-nguyen** [comment]
@maxschulz-COL @petar-qb - do you guys want to take a final look? :)

Also any ideas why the docs are failing on some links I didn't add? 🤔  These are pre-existing links in the data-manager and actions user guide and I've just checked and they still work, so not sure why it's failing.

File: vizro-core/docs/pages/user-guides/graph.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/graph.md:72]:
```suggestion
If you want to enhance your `Graph` with a title, header, or footer to provide additional context or descriptions,
```

File: vizro-core/docs/pages/user-guides/table.md
Comment (stichbury) [inline vizro-core/docs/pages/user-guides/table.md:535]:
```suggestion
To enhance your `Table` or `AgGrid` with a title, header, or footer to provide additional context or
```

**Review by stichbury — APPROVED** [review]
LGTM, only a couple of very minor comments 🌟

File: vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py
Comment (petar-qb) [inline vizro-core/tests/unit/vizro/models/_components/test_ag_grid.py:1]: May you just remove `filter_interaction_action` from `test_pre_build_actions_underlying_ag_grid_id` and `test_pre_build_actions_underlying_table_id`? 

The underlying table ID no longer depends on the actions attached to main model (vm.AgGrid/vm.Table). We should have removed this long time ago, but somehow it slipped through.

**Review by petar-qb — APPROVED** [review]
Amazing work! This looks really good to me now. 🎖️

File: vizro-core/src/vizro/models/_components/graph.py
Comment (maxschulz-COL) [reply vizro-core/src/vizro/models/_components/graph.py]: Ah weird, did you properly dedent (either manually, by putting the text to the very left), or by using
```py
from textwrap import dedent
print(dedent("""
	Text for which you want to remove
    leading whitespace when printed
    """))
```

Very nice tool :P

**Comment by maxschulz-COL** [comment]
> @maxschulz-COL @petar-qb - do you guys want to take a final look? :)
> 
> @stichbury - If you have some time to take a look at the docs here, that would also be great! 🙇‍♀️

Too late to the party, but look great 😍

File: vizro-core/docs/pages/user-guides/titles-header-footer.md
Comment (huong-li-nguyen) [reply vizro-core/docs/pages/user-guides/titles-header-footer.md:1]: Ah @petar-qb - I understood what you meant now. We do not need to update the  `requirements.txt` for all py.cafe, because these should run fine with the previous versions and they do not use any of the new features. 

What we update is our `mkdocs.yml` such that our pycafe examples there run on the newest version. I think that's what @maxschulz-COL also referred to, there was already a step in the release process :)

File: vizro-core/src/vizro/models/_components/graph.py
Comment (petar-qb) [reply vizro-core/src/vizro/models/_components/graph.py:1]: FYI: @huong-li-nguyen @antonymilne 

Here's an example [Issue #727](https://github.com/mckinsey/vizro/issues/727) where adding a title, along with a header and footer, would be beneficial. 

Do you think adding a title/header/footer is a better idea than directing users to write their own html elements in the figure function? In that case, users should add the appropriate classname to these html components so they react on the theme change.

**Pull request merged**
